tapgarden/caretaker.go

```diff
@@ -244,22 +262,20 @@ func (b *BatchCaretaker) advanceStateUntil(currentState,
 			return 0, fmt.Errorf("BatchCaretaker(%x), shutting "+
 				"down", b.batchKey[:])
 
+		// If the batch was cancellable, the finalState of the cancel
+		// response will be non-nil. If the cancellation failed, that
+		// error will be handled by the planter. At this point, the
+		// caretaker should always shut down gracefully.
 		case <-b.cfg.CancelReqChan:
-			cancelResp := b.Cancel()
-			b.cfg.CancelRespChan <- cancelResp
-
-			// TODO(jhb): Use concrete error types for caretaker
-			// shutdown cases
-			// If the batch was cancellable, the finalState of the
-			// cancel response will be non-nil. If the cancellation
-			// failed, that error will be handled by the planter.
-			// At this point, the caretaker should always shut down
-			// gracefully.
-			if cancelResp.finalState != nil {
+			cancelErr := b.Cancel()
+			if cancelErr == nil {
```

Comment:
Shouldn't we turn around the logic around the error being nil vs. non-nil? I see this replicates the previous logic, but I would expect a `cancelErr != nil` here rather.

Reply:
I agree the use of `Cancel()` is a bit confusing - we send the error from the attempt to cancel back to the planter in `Cancel()` and this error is only about if cancellation was attempted in `Cancel()`.

So if cancellation was possible and attempted, this err is nil and we shut down. If cancellation was not possible, we return an error and keep running.

I'll add detail on that for the doc for `Cancel()`.